src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,550 @@
+        q_mul: config.q_mul_hi,
+        z: config.bits,
+        x_a: config.extras[0],
+        lambda: config.lambda,
+    };
+    let lo_columns = IncompleteColumns {
+        q_mul: config.q_mul_lo,
+        z: config.extras[1],
+        x_a: config.extras[2],
+        lambda: (config.extras[3], config.extras[4]),
+    };
+
+    // Decompose the scalar bitwise (big-endian bit order).
+    let k_bits = decompose_scalar::<C>(scalar.value.unwrap());
+
+    // Bits used in incomplete addition. k_{254} to k_{4} inclusive
+    let incomplete_range = 0..(C::Scalar::NUM_BITS as usize - 1 - NUM_COMPLETE_BITS);
+    let k_incomplete = &k_bits[incomplete_range];
+    let k_incomplete_hi = &k_incomplete[..k_incomplete.len() / 2];
+    let k_incomplete_lo = &k_incomplete[k_incomplete.len() / 2..];
+
+    // Bits used in complete addition. k_{3} to k_{1} inclusive
+    // The LSB k_{0} is handled separately.
+    let complete_range =
+        (C::Scalar::NUM_BITS as usize - 1 - NUM_COMPLETE_BITS)..(C::Scalar::NUM_BITS as usize - 1);
+    let k_complete = &k_bits[complete_range.clone()];
+
+    // Initialize the accumulator a [2]base
+    let acc = double::assign_region(&base, offset, region, config.clone())?;
+    // Initialize the running sum for scalar decomposition to zero
+    let z_val = C::Base::zero();
+    let z_cell = region.assign_advice(|| "initial z", hi_columns.z, offset + 1, || Ok(z_val))?;
+    let z = CellValue::new(z_cell, Some(z_val));
+
+    // Double-and-add (incomplete addition) for the `hi` half of the scalar decomposition
+    let (x, y_a, z) = add_incomplete::<C>(
+        region,
+        &base,
+        config.clone(),
+        offset + 1,
+        hi_columns,
+        k_incomplete_hi,
+        (X(acc.x.clone()), Y(acc.y.value), ZValue(z)),
+    )?;
+
+    // Double-and-add (incomplete addition) for the `lo` half of the scalar decomposition
+    let (x, y_a, z) = add_incomplete::<C>(
+        region,
+        &base,
+        config.clone(),
+        offset + 1,
+        lo_columns,
+        k_incomplete_lo,
+        (x, y_a, z),
+    )?;
+
+    // Move from incomplete addition to complete addition
+    let mut acc = {
+        let y_a_col = config.extras[1];
+        let row = k_incomplete_lo.len() + 2;
+
+        let y_a_cell = region.assign_advice(
+            || "y_a",
+            y_a_col,
+            row + offset,
+            || y_a.ok_or(Error::SynthesisError),
+        )?;
+        util::assign_and_constrain(
+            region,
+            || "Copy z from incomplete to complete",
+            config.bits.into(),
+            row + offset,
+            &z,
+            &config.perm,
+        )?;
+        EccPoint {
+            x: x.0,
+            y: CellValue::<C::Base>::new(y_a_cell, *y_a),
+        }
+    };
+
+    let mut z_val = z.value;
+    // Complete addition
+    for (iter, k) in k_complete.iter().enumerate() {
+        // Each iteration uses 4 rows (two complete additions)
```

Comment:
I believe we can now do a complete addition in one row.

Reply:
This has now been implemented.